File: internal/record/record_test.go
package record

import "testing"

func TestRecordEncodeDecodeRoundTrip(t *testing.T) {
	original := Record{
		Key:       []byte("user:1"),
		Value:     []byte("alice"),
		Timestamp: 123456789,
		IsDeleted: false,
	}

	encoded, err := original.Encode()
	if err != nil {
		t.Fatalf("encode failed: %v", err)
	}

	decoded, err := Decode(encoded)
	if err != nil {
		t.Fatalf("decode failed: %v", err)
	}

	if string(decoded.Key) != string(original.Key) {
		t.Fatalf("key mismatch: got %q want %q", decoded.Key, original.Key)
	}
	if string(decoded.Value) != string(original.Value) {
		t.Fatalf("value mismatch: got %q want %q", decoded.Value, original.Value)
	}
	if decoded.Timestamp != original.Timestamp {
		t.Fatalf("timestamp mismatch: got %d want %d", decoded.Timestamp, original.Timestamp)
	}
	if decoded.IsDeleted != original.IsDeleted {
		t.Fatalf("deleted flag mismatch: got %v want %v", decoded.IsDeleted, original.IsDeleted)
	}
}

func TestDeleteRecordRoundTrip(t *testing.T) {
	original := NewDelete([]byte("user:2"), 999)

	encoded, err := original.Encode()
	if err != nil {
		t.Fatalf("encode failed: %v", err)
	}

	decoded, err := Decode(encoded)
	if err != nil {
		t.Fatalf("decode failed: %v", err)
	}

	if !decoded.IsDeleted {
		t.Fatalf("expected tombstone record")
	}
	if string(decoded.Key) != "user:2" {
		t.Fatalf("key mismatch: got %q", decoded.Key)
	}
}

File: internal/record/record_test.cpp
#include "record.h"
#include <gtest/gtest.h>
#include <string>
#include <vector>

using namespace bigdb::record;

TEST(RecordTest, EncodeDecodeRoundTrip) {
    std::string key_str = "user:1";
    std::string value_str = "alice";
    std::vector<uint8_t> key(key_str.begin(), key_str.end());
    std::vector<uint8_t> value(value_str.begin(), value_str.end());
    int64_t timestamp = 123456789;

    Record original = Record::new_put(key, value, timestamp);

    std::vector<uint8_t> encoded = original.encode();
    Record decoded = Record::decode(encoded);

    EXPECT_EQ(decoded.key, original.key);
    EXPECT_EQ(decoded.value, original.value);
    EXPECT_EQ(decoded.timestamp, original.timestamp);
    EXPECT_EQ(decoded.is_deleted, original.is_deleted);
}

TEST(RecordTest, DeleteRecordRoundTrip) {
    std::string key_str = "user:2";
    std::vector<uint8_t> key(key_str.begin(), key_str.end());
    int64_t timestamp = 999;

    Record original = Record::new_delete(key, timestamp);

    std::vector<uint8_t> encoded = original.encode();
    Record decoded = Record::decode(encoded);

    EXPECT_TRUE(decoded.is_deleted);
    EXPECT_EQ(decoded.key, original.key);
    EXPECT_TRUE(decoded.value.empty());
    EXPECT_EQ(decoded.timestamp, original.timestamp);
}
